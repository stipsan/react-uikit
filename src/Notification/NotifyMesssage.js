import { Component, PropTypes } from 'react'
import cx from 'classnames'
import Icon from '../Icon'

export default class NotifyMesssage extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    icon: PropTypes.string,
    isSticky: PropTypes.bool,
    timeout: PropTypes.number,
    type: PropTypes.oneOf(['info', 'success', 'danger', 'warning', 'primary', false]),
    onClick: PropTypes.func, // eslint-disable-line react/require-default-props
  }

  static defaultProps = {
    icon: '',
    isSticky: false,
    timeout: 3000,
    type: false,
  }

  state = {
    isOpen: false,
  }

  componentDidMount() {
    setTimeout(this.openNotification, 100)

    if (this.props.timeout && !this.props.isSticky) {
      setTimeout(this.handleClose, this.props.timeout)
    }
  }

  openNotification = () => this.setState({ isOpen: true })

  handleClose = () => {
    this.setState({ isOpen: false }, () => setTimeout(() => this.setState({ isClosed: true }), 300))
  }

  render() {
    const { children, type, icon, onClick } = this.props
    const className = cx('uk-notification-message', {
      'uk-notification-message-info': type === 'info',
      'uk-notification-message-success': type === 'success',
      'uk-notification-message-danger': type === 'danger',
      'uk-notification-message-warning': type === 'warning',
      'uk-notification-message-primary': type === 'primary',
    })
    const styles = [
      { marginTop: '-64px' },
      { marginTop: '0px' },
    ]

    if (this.state.isClosed) {
      return false
    }

    const body = icon ? (
      <div onClick={onClick}>
        <Icon icon={icon} />
        &nbsp;
        {children}
      </div>
    ) : (
      <div onClick={onClick}>
        {children}
      </div>
  )

    return (
      <div
        className={className}
        style={{ overflow: 'hidden', transition: 'margin ease-out 300ms', ...styles[this.state.isOpen ? 1 : 0] }}
      >
        <a className="uk-notification-close uk-close uk-icon" onClick={this.handleClose} >
          <Icon icon="close" />
        </a>
        {body}
      </div>
    )
  }
}
