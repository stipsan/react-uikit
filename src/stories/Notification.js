import { storiesOf, action } from '@kadira/storybook'
import { Notification, Button } from 'uikit-react'
import { PureComponent, PropTypes } from 'react'

class NotificationsContainer extends PureComponent {
  static propTypes = {
    position: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    icon: PropTypes.string, // eslint-disable-line react/require-default-props
    isSticky: PropTypes.bool, // eslint-disable-line react/require-default-props
    timeout: PropTypes.number, // eslint-disable-line react/require-default-props
    onClick: PropTypes.func, // eslint-disable-line react/require-default-props
  }

  static defaultProps = {
    isSticky: false,
    position: 'top-center',
    type: 'info',
  }

  state = {
    notifications: [],
    count: 1,
  }

  handleClick = (event) => {
    event.preventDefault()
    this.setState({
      notifications: [
        {
          id: this.state.count,
          type: this.props.type,
          icon: this.props.icon,
          message: `Notification #${this.state.count}`,
          isSticky: this.props.isSticky,
          timeout: this.props.timeout,
          onClick: this.props.onClick ? this.props.onClick : () => {},
        },
        ...this.state.notifications],
      count: this.state.count + 1,
    })
  }

  render() {
    const { notifications } = this.state
    const { position, type } = this.props
    return (<div>
      <Button
        className={`uk-margin-bottom uk-button uk-button-${type}`}
        onClick={this.handleClick}
      >
        {`${type}-${position}`}
        {this.props.isSticky ? '- isSticky' : ''}
        {this.props.icon ? '- icon' : ''}
        {this.props.onClick ? '- click' : ''}
        {this.props.timeout ? '- timeout' : ''}
      </Button>
      <Notification
        notifications={notifications}
        position={position}
      />
    </div>)
  }
}

Notification.displayName = 'Notification'

storiesOf('Notification', module)
  .addWithInfo('Basic Usage', '', () => (
    <div className="uk-margin-bottom">
      <NotificationsContainer type="info" />
    </div>
  ), { header: false, inline: true, propTables: [Notification] })
  .addWithInfo('Position', '', () => (
    <div className="uk-margin-bottom">
      <NotificationsContainer position="top-left" type="info" />
      <NotificationsContainer position="top-center" type="info" />
      <NotificationsContainer position="top-right" type="info" />
      <NotificationsContainer position="bottom-left" type="info" />
      <NotificationsContainer position="bottom-center" type="info" />
      <NotificationsContainer position="bottom-right" type="info" />
    </div>
  ), { header: false, inline: true, propTables: [Notification] })
  .addWithInfo('Style', '', () => (
    <div className="uk-margin-bottom">
      <NotificationsContainer position="bottom-center" type="info" />
      <NotificationsContainer position="top-left" type="primary" />
      <NotificationsContainer position="top-center" type="success" />
      <NotificationsContainer position="top-right" type="warning" />
      <NotificationsContainer position="bottom-left" type="danger" />
    </div>
  ), { header: false, inline: true, propTables: [Notification] })
  .addWithInfo('Component options', '', () => (
    <div className="uk-margin-bottom">
      <NotificationsContainer icon="warning" position="bottom-left" timeout={10000} type="warning" />
      <NotificationsContainer icon="warning" position="top-right" timeout={5000} type="danger" onClick={action('handleClick')} />
      <NotificationsContainer position="top-center" type="success" />
      <NotificationsContainer position="top-right" type="warning" />
      <NotificationsContainer icon="danger" position="top-left" type="danger" />
    </div>
  ), { header: false, inline: true, propTables: [Notification] })
