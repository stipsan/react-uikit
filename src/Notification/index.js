import { Component, PropTypes } from 'react'
import Portal from 'react-portal'
import cx from 'classnames'

import NotifyMesssage from './NotifyMesssage'

export default class Notification extends Component {
  static propTypes = {
    notifications: PropTypes.arrayOf(PropTypes.object).isRequired,
    position: PropTypes.string,
  }

  static defaultProps = {
    type: 'primary',
    className: 'uk-notification',
    position: 'top-center',
    notifications: [],
  }

  state = {
    isOpened: false,
  }

  handleClick = () => {
    this.setState({
      isOpened: true,
    })
  }

  render() {
    const {
      position,
      notifications,
    } = this.props

    const className = cx('uk-notification', {
      'uk-notification-top-center': position === 'top-center',
      'uk-notification-top-left': position === 'top-left',
      'uk-notification-top-right': position === 'top-right',
      'uk-notification-bottom-center': position === 'bottom-center',
      'uk-notification-bottom-left': position === 'bottom-left',
      'uk-notification-bottom-right': position === 'bottom-right',
    })
    return (
      <Portal
        isOpened={notifications.length > 0}
        onClose={this.handleClick}
      >
        <div className={className}>
          {notifications.map(({ type, message, id, icon, isSticky, onClick, timeout }) => (
            <NotifyMesssage
              icon={icon}
              isSticky={isSticky}
              key={id}
              timeout={timeout}
              type={type}
              onClick={onClick}
            >
              {message}
            </NotifyMesssage>
          ))}
        </div>
      </Portal>
    )
  }
}
