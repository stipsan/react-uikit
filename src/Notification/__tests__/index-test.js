import renderer from 'react-test-renderer'

import Notification from '../index'

jest.mock('react-portal', () => 'Portal')

describe('Notification', () => {
  const defaultProps = {
    notifications: [{ id: 1, type: 'warning', message: 'Notification #1', icon: 'check' }],
    position: 'top-center',
  }
  const component = renderer.create(
    <Notification {...defaultProps} />
  )
  it('should render correctly', () => {
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should hanlde click correctly', () => {
    const instance = component.getInstance()
    instance.handleClick()
    expect(instance.state.isOpened).toBe(true)
  })
})
