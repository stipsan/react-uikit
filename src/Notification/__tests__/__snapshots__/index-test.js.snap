// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification should render correctly 1`] = `
<Portal
  isOpened={true}
  onClose={[Function]}
>
  <div
    className="uk-notification uk-notification-top-center"
  >
    <div
      className="uk-notification-message uk-notification-message-warning"
      style={
        Object {
          "marginTop": "-64px",
          "overflow": "hidden",
          "transition": "margin ease-out 300ms",
        }
      }
    >
      <a
        className="uk-close"
        onClick={[Function]}
      />
      <div
        onClick={undefined}
      >
        <span>
          <i
            className="uk-icon-check"
          />
          Â 
        </span>
        Notification #1
      </div>
    </div>
  </div>
</Portal>
`;
